Add additional C++ Example for AM2302-Sensor

hasenradball opened the pull request:
# Add a C++ Example for AM2302-Sensor.

- usage of class instantiation
- based on [AM2302-Sensor Lib](https://github.com/hasenradball/AM2302-Sensor)

hasenradball committed:
add c++ example for AM2302_sensor

hasenradball commented:
@lurch Hi,

added in short a cpp example for the AM2302-sensor.
It is possible to add to the pico examples?

lurch commented:
As I might have mentioned before, I only provide "(hopefully) helpful comments", it's not my decision whether to merge PRs in `pico-sdk` or `pico-examples`. But my first helpful comment would be that you need to target your PR against `develop` instead of `master` :wink:

hasenradball commented:
@lurch Thanks for the hint.
You are totally right.

:-)

peterharperuk commented:
How is this different from the DHT22 code you've added? Or are you adding this as an example of how to use C++?

hasenradball commented:
@peterharperuk 
Hi Peter, the idea was to add this as C++ Example.
I got the info in the morning that cpp in gerneral is possible… using classes and instatiation and all this stuff.
Then I tried it directly.
It makes the main file a little cleaner.
Could be an example how it looks like in cpp style.
One thing is different here.
When call method begin() there is a check if sensor is connected.

And maybee here could also the check placed to distinguish between DHT11 or DHT22.

Best regards
Frank

hasenradball commented:
@peterharperuk 
One other thing ist also improved.
If you poll the sensor too often which means < 2 s .
Then status is -3 and the read function is not called. 

So code is slightly improved.
please try it.

hasenradball committed:
remame and comment added for readfrequency variable

hasenradball committed:
rework for LED PIN

lurch commented on gpio/AM2302_sensor/AM2302.cpp:
```diff
@@ -8,7 +8,7 @@
 #include "AM2302-Sensor.hpp"
 
-const uint LED = 13U;
+const uint LED_PIN = 13U;
```
Maybe you'd like to make use of the `PICO_DEFAULT_LED_PIN` define, so that your example doesn't require an external LED? https://github.com/raspberrypi/pico-sdk/blob/master/src/boards/include/boards/pico.h#L33
(see also e.g. https://github.com/raspberrypi/pico-examples/blob/master/blink/blink.c )

hasenradball committed:
use default LED PIN

hasenradball commented:
@lurch can you please give a review for this PR.

lurch commented:
> As I might have mentioned before, I only provide "(hopefully) helpful comments", it's not my decision whether to merge PRs in `pico-sdk` or `pico-examples`.

hasenradball commented:
@lurch I thought so because you are mentioned as reviewer, sorry for the missunderstanding.

lurch commented:
I think GitHib just treats _any_ comment on the code as a "review" :facepalm:

hasenradball committed:
improve error handling; and use Sensor Array as example

hasenradball commented:
@hasenradball 

Added the following features:
- better human readable error output
- example with sensor array

Added screenshot how example looks: 

![20240425_16h59m19s_grim](https://github.com/raspberrypi/pico-examples/assets/26808969/26faa763-e028-4c2d-83ff-695d76d4ca67)

lurch commented on gpio/AM2302_sensor/AM2302-Sensor.cpp:
```diff
@@ -0,0 +1,252 @@
+      sleep_ms(1U);
+   }
+   auto status{read()};
+   _millis_last_read = time_us_64();
```
`time_us_64` returns the time in microseconds https://www.raspberrypi.com/documentation/pico-sdk/hardware.html#ga7bbcd996dbd3dd4adc03a80272fb2fc3 so calling the variable `_millis_last_read` might be confusing to whoever reads this code in future? :wink:

hasenradball replied:
sure youre right,
I should adapt this
👍

lurch commented on gpio/AM2302_sensor/AM2302-Sensor.cpp:
```diff
@@ -0,0 +1,252 @@
+/*
+*	AM2302-Sensor.cpp
+*
+*	Author: Frank Häfele
```
Could you add some licensing-information please?

hasenradball replied:
@lurch 
any prefererations?

lurch replied:
As you're the author, it's obviously up to you. But if you choose the same licence as all of our other examples, that's probably easiest.

hasenradball replied:
@lurch fixed, hope it is fine.

lurch replied:
For the avoidance of doubt, it's probably best to add the `License-Identifier` to each of the header-blocks in which you've listed yourself as the Author?

hasenradball committed:
fix review findings

hasenradball committed:
updated file Information and licence

hasenradball committed:
make get_sensorState() static
